File: aetherctl/src/openspec/state.ts
import { createHash, randomUUID } from "node:crypto";
import { mkdir, open, readFile, rename, stat, unlink, utimes, type FileHandle } from "node:fs/promises";
import { dirname, join } from "node:path";
import { safePath, sanitizePathComponent, STALE_LOCK_AGE_MS } from "./security.js";

export type OpenSpecStatus =
  | "PLANNING"
  | "WAIT_APPROVAL"
  | "PRECHECK"
  | "EXECUTING"
  | "VERIFYING"
  | "CANCEL_REQUESTED"
  | "PAUSED_UNEXPECTED"
  | "DONE"
  | "CANCELLED"
  | "FAILED";

export type OpenSpecStateV1 = {
  schemaVersion: 1;
  changeId: string;
  repoPath: string;
  status: OpenSpecStatus;
  createdAt: string;
  updatedAt: string;

  approval: {
    approved: boolean;
    approvedAt: string | null;
    approvedBy: string | null;
    approvedPlanDigest: string | null;
  };

  plan: {
    planDigest: string | null;
    headSha: string | null;
  };

  execution: {
    runId: string | null;
    attempt: number;
    preflightPassed: boolean;
    currentTask: string | null;
    currentStepId: string | null;
    startedAt: string | null;
    lastHeartbeatAt: string | null;
  };

  loop: {
    lastTickAt: string | null;
    ticks: number;
    maxTicks: number;
    maxRunSeconds: number;
  };

  blocked: {
    reason: string | null;
    needs: unknown | null;
    pause: {
      kind: string | null;
      request: string | null;
      unblockCommand: string | null;
      resumeCondition: string | null;
      userInput: string | null;
      requestedAt: string | null;
      resolvedAt: string | null;
    };
  };

  error: {
    code: string | null;
    message: string | null;
    detail: unknown | null;
    retryable: boolean | null;
    firstSeenAt: string | null;
    lastSeenAt: string | null;
  };
};

export function statePathFor(repoPath: string, changeId: string): string {
  // Validate and sanitize changeId to prevent path traversal
  const sanitizedChangeId = sanitizePathComponent(changeId);
  return safePath(repoPath, "openspec", "changes", sanitizedChangeId, "state.json");
}

export function progressLogPathFor(repoPath: string, changeId: string): string {
  // Validate and sanitize changeId to prevent path traversal
  const sanitizedChangeId = sanitizePathComponent(changeId);
  return safePath(repoPath, "openspec", "changes", sanitizedChangeId, "progress.log");
}

export async function loadState(repoPath: string, changeId: string): Promise<OpenSpecStateV1> {
  const p = statePathFor(repoPath, changeId);
  const raw = await readFile(p, "utf8");
  const json = JSON.parse(raw) as OpenSpecStateV1;
  if (json.schemaVersion !== 1) throw new Error(`Unsupported state schemaVersion: ${(json as any).schemaVersion}`);
  return json;
}

export async function ensureState(repoPath: string, changeId: string): Promise<OpenSpecStateV1> {
  const p = statePathFor(repoPath, changeId);
  try {
    const s = await stat(p);
    if (s.isFile()) return await loadState(repoPath, changeId);
  } catch {
    // ignore
  }

  const now = new Date().toISOString();
  const init: OpenSpecStateV1 = {
    schemaVersion: 1,
    changeId,
    repoPath,
    status: "PLANNING",
    createdAt: now,
    updatedAt: now,
    approval: { approved: false, approvedAt: null, approvedBy: null, approvedPlanDigest: null },
    plan: { planDigest: null, headSha: null },
    execution: {
      runId: null,
      attempt: 0,
      preflightPassed: false,
      currentTask: null,
      currentStepId: null,
      startedAt: null,
      lastHeartbeatAt: null
    },
    loop: { lastTickAt: null, ticks: 0, maxTicks: 200, maxRunSeconds: 1800 },
    blocked: {
      reason: null,
      needs: null,
      pause: {
        kind: null,
        request: null,
        unblockCommand: null,
        resumeCondition: null,
        userInput: null,
        requestedAt: null,
        resolvedAt: null
      }
    },
    error: { code: null, message: null, detail: null, retryable: null, firstSeenAt: null, lastSeenAt: null }
  };

  await atomicWriteJson(p, init);
  return init;
}

export async function atomicWriteJson(path: string, data: unknown): Promise<void> {
  await mkdir(dirname(path), { recursive: true });
  const tmp = `${path}.tmp.${process.pid}.${Date.now()}`;
  const payload = JSON.stringify(data, null, 2) + "\n";

  let fh: FileHandle | null = null;
  try {
    fh = await open(tmp, "w");
    await fh.writeFile(payload, "utf8");
    await fh.sync();
    await fh.close();
    fh = null;

    await rename(tmp, path);
  } catch (e) {
    // Clean up temporary file on error
    try {
      if (fh) await fh.close();
      await unlink(tmp);
    } catch {
      // Ignore cleanup errors
    }
    throw e;
  }
}

export async function updateState(
  repoPath: string,
  changeId: string,
  fn: (prev: OpenSpecStateV1) => OpenSpecStateV1
): Promise<OpenSpecStateV1> {
  const statePath = statePathFor(repoPath, changeId);
  const lockPath = `${statePath}.lock`;

  const release = await acquireFsLock(lockPath, 5000);
  try {
    const prev = await ensureState(repoPath, changeId);
    const next = fn(prev);
    next.updatedAt = new Date().toISOString();
    await atomicWriteJson(statePath, next);
    return next;
  } finally {
    await release();
  }
}

async function acquireFsLock(lockPath: string, timeoutMs: number): Promise<() => Promise<void>> {
  await mkdir(dirname(lockPath), { recursive: true });
  const started = Date.now();
  const pid = process.pid;
  const maxRetries = Math.ceil(timeoutMs / 50);
  let retries = 0;

  while (retries < maxRetries) {
    try {
      // Check for and clean up stale locks
      try {
        const stats = await stat(lockPath);
        const age = Date.now() - stats.mtimeMs;
        if (age > STALE_LOCK_AGE_MS) {
          console.warn(`[openspec] Removing stale lock (age: ${Math.round(age / 1000)}s): ${lockPath}`);
          await unlink(lockPath);
        } else {
          // Lock exists and is not stale, check if owner process is alive
          const content = await readFile(lockPath, "utf8").catch(() => "");
          const [lockPid] = content.split("\n");
          if (lockPid && lockPid !== String(pid)) {
            // Try to check if process exists (Unix-like systems)
            try {
              process.kill(Number(lockPid), 0); // Signal 0 checks existence
              // Process exists, lock is valid
            } catch {
              // Process doesn't exist, lock is orphaned
              console.warn(`[openspec] Removing orphaned lock (pid: ${lockPid}): ${lockPath}`);
              await unlink(lockPath);
            }
          }
        }
      } catch {
        // Lock file doesn't exist or error reading, continue to acquire
      }

      // Atomically create lock file with PID and timestamp
      const fh = await open(lockPath, "wx");
      try {
        await fh.writeFile(`${pid}\n${Date.now()}\n`, "utf8");
        await fh.sync();
      } finally {
        await fh.close();
      }

      // Successfully acquired lock, return release function
      return async () => {
        try {
          // Verify lock still belongs to us before releasing
          const content = await readFile(lockPath, "utf8");
          const [lockPid] = content.split("\n");
          if (lockPid === String(pid)) {
            await unlink(lockPath);
          } else {
            console.warn(`[openspec] Lock was taken by another process (current: ${lockPid}, ours: ${pid}): ${lockPath}`);
          }
        } catch (e) {
          console.error(`[openspec] Failed to release lock: ${e}`);
          throw e; // Re-throw to make lock release failures visible
        }
      };
    } catch (e: any) {
      if (e?.code !== "EEXIST") throw e;

      retries++;
      if (Date.now() - started > timeoutMs || retries >= maxRetries) {
        throw new Error(`Timed out acquiring state lock after ${retries} retries: ${lockPath}`);
      }

      // Exponential backoff with jitter
      const backoff = Math.min(50 * Math.pow(1.5, retries), 1000);
      const jitter = Math.random() * backoff * 0.1;
      await new Promise((r) => setTimeout(r, backoff + jitter));
    }
  }

  throw new Error(`Failed to acquire lock after ${retries} retries: ${lockPath}`);
}

export async function computePlanDigest(repoPath: string, changeId: string): Promise<string> {
  // Sanitize changeId to prevent path traversal
  const sanitizedChangeId = sanitizePathComponent(changeId);

  const parts: Array<[string, string]> = [];
  const files = [
    safePath(repoPath, "openspec", "config.yaml"),
    safePath(repoPath, "openspec", "changes", sanitizedChangeId, "proposal.md"),
    safePath(repoPath, "openspec", "changes", sanitizedChangeId, "design.md"),
    safePath(repoPath, "openspec", "changes", sanitizedChangeId, "tasks.md")
  ];

  for (const f of files) {
    let content = "";
    try {
      content = await readFile(f, "utf8");
    } catch {
      content = "";
    }
    parts.push([f, normalizeText(content)]);
  }

  const h = createHash("sha256");
  for (const [f, c] of parts) {
    h.update(f);
    h.update("\n");
    h.update(c);
    h.update("\n\n");
  }
  return h.digest("hex");
}

export function newRunId(): string {
  return randomUUID();
}

function normalizeText(s: string): string {
  return s.replace(/\r\n/g, "\n").trimEnd() + "\n";
}

File: nanobot/agent/taskmode/state.py
"""Task-Mode state management — atomic read/write of state.json."""

from __future__ import annotations

import hashlib
import json
import os
import tempfile
import uuid
from dataclasses import asdict, dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any, Callable

from loguru import logger


# ---------------------------------------------------------------------------
# Status constants
# ---------------------------------------------------------------------------

PLANNING = "PLANNING"
WAIT_APPROVAL = "WAIT_APPROVAL"
PRECHECK = "PRECHECK"
EXECUTING = "EXECUTING"
VERIFYING = "VERIFYING"
CANCEL_REQUESTED = "CANCEL_REQUESTED"
PAUSED_UNEXPECTED = "PAUSED_UNEXPECTED"
DONE = "DONE"
CANCELLED = "CANCELLED"
FAILED = "FAILED"

TERMINAL_STATUSES = {DONE, CANCELLED, FAILED}
RUNNING_STATUSES = {PRECHECK, EXECUTING, VERIFYING}


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------

@dataclass
class ApprovalInfo:
    approved: bool = False
    approved_at: str | None = None
    approved_by: str | None = None
    approved_plan_digest: str | None = None


@dataclass
class PlanInfo:
    plan_digest: str | None = None
    head_sha: str | None = None


@dataclass
class ExecutionInfo:
    run_id: str | None = None
    attempt: int = 0
    preflight_passed: bool = False
    current_task: str | None = None
    current_step_id: str | None = None
    started_at: str | None = None
    last_heartbeat_at: str | None = None


@dataclass
class LoopInfo:
    last_tick_at: str | None = None
    ticks: int = 0
    max_ticks: int = 200
    max_run_seconds: int = 1800


@dataclass
class PauseInfo:
    kind: str | None = None
    request: str | None = None
    unblock_command: str | None = None
    resume_condition: str | None = None
    user_input: str | None = None
    requested_at: str | None = None
    resolved_at: str | None = None


@dataclass
class BlockedInfo:
    reason: str | None = None
    needs: Any | None = None
    pause: PauseInfo = field(default_factory=PauseInfo)


@dataclass
class ErrorInfo:
    code: str | None = None
    message: str | None = None
    detail: Any | None = None
    retryable: bool | None = None
    first_seen_at: str | None = None
    last_seen_at: str | None = None


@dataclass
class TaskState:
    """OpenSpec change state, mirroring the TypeScript ``OpenSpecStateV1``."""

    schema_version: int = 1
    change_id: str = ""
    repo_path: str = ""
    status: str = PLANNING
    created_at: str = ""
    updated_at: str = ""
    approval: ApprovalInfo = field(default_factory=ApprovalInfo)
    plan: PlanInfo = field(default_factory=PlanInfo)
    execution: ExecutionInfo = field(default_factory=ExecutionInfo)
    loop: LoopInfo = field(default_factory=LoopInfo)
    blocked: BlockedInfo = field(default_factory=BlockedInfo)
    error: ErrorInfo = field(default_factory=ErrorInfo)


# ---------------------------------------------------------------------------
# Serialisation helpers (snake_case <-> camelCase)
# ---------------------------------------------------------------------------

def _to_camel(name: str) -> str:
    parts = name.split("_")
    return parts[0] + "".join(p.capitalize() for p in parts[1:])


def _to_snake(name: str) -> str:
    import re
    return re.sub(r"(?<=[a-z0-9])([A-Z])", r"_\1", name).lower()


def _state_to_dict(state: TaskState) -> dict[str, Any]:
    """Serialise *state* to a JSON-friendly dict with **camelCase** keys."""

    def _convert(obj: Any) -> Any:
        if isinstance(obj, dict):
            return {_to_camel(k): _convert(v) for k, v in obj.items()}
        if isinstance(obj, list):
            return [_convert(i) for i in obj]
        return obj

    return _convert(asdict(state))


def _dict_to_state(data: dict[str, Any]) -> TaskState:
    """Deserialise a camelCase dict into a ``TaskState``."""

    def _snake_dict(d: dict) -> dict:
        return {_to_snake(k): _snake_dict(v) if isinstance(v, dict) else v for k, v in d.items()}

    flat = _snake_dict(data)

    return TaskState(
        schema_version=flat.get("schema_version", 1),
        change_id=flat.get("change_id", ""),
        repo_path=flat.get("repo_path", ""),
        status=flat.get("status", PLANNING),
        created_at=flat.get("created_at", ""),
        updated_at=flat.get("updated_at", ""),
        approval=ApprovalInfo(**{k: v for k, v in flat.get("approval", {}).items() if k in ApprovalInfo.__dataclass_fields__}),
        plan=PlanInfo(**{k: v for k, v in flat.get("plan", {}).items() if k in PlanInfo.__dataclass_fields__}),
        execution=ExecutionInfo(**{k: v for k, v in flat.get("execution", {}).items() if k in ExecutionInfo.__dataclass_fields__}),
        loop=LoopInfo(**{k: v for k, v in flat.get("loop", {}).items() if k in LoopInfo.__dataclass_fields__}),
        blocked=_parse_blocked(flat.get("blocked", {})),
        error=ErrorInfo(**{k: v for k, v in flat.get("error", {}).items() if k in ErrorInfo.__dataclass_fields__}),
    )


def _parse_blocked(d: dict) -> BlockedInfo:
    pause_raw = d.get("pause", {})
    pause = PauseInfo(**{k: v for k, v in pause_raw.items() if k in PauseInfo.__dataclass_fields__})
    return BlockedInfo(
        reason=d.get("reason"),
        needs=d.get("needs"),
        pause=pause,
    )


# ---------------------------------------------------------------------------
# StateManager
# ---------------------------------------------------------------------------

def _sanitize_path_component(name: str) -> str:
    """Prevent path traversal in change IDs."""
    name = name.replace("/", "_").replace("\\", "_").replace("..", "_")
    return name


def new_run_id() -> str:
    return str(uuid.uuid4())


class StateManager:
    """Atomic read/write for ``state.json`` inside an OpenSpec change directory."""

    # ------ path helpers ------

    @staticmethod
    def change_dir(workspace: Path, change_id: str) -> Path:
        safe = _sanitize_path_component(change_id)
        return workspace / "openspec" / "changes" / safe

    @staticmethod
    def state_path(change_dir: Path) -> Path:
        return change_dir / "state.json"

    @staticmethod
    def progress_log_path(change_dir: Path) -> Path:
        return change_dir / "progress.log"

    # ------ read / write ------

    def read(self, change_dir: Path) -> TaskState:
        """Load state from ``state.json``.  Raises if missing."""
        path = self.state_path(change_dir)
        raw = path.read_text(encoding="utf-8")
        data = json.loads(raw)
        if data.get("schemaVersion") != 1:
            raise ValueError(f"Unsupported state schemaVersion: {data.get('schemaVersion')}")
        return _dict_to_state(data)

    def read_or_init(self, change_dir: Path, change_id: str, repo_path: str = "") -> TaskState:
        """Load existing state or initialise a fresh one."""
        path = self.state_path(change_dir)
        if path.exists():
            return self.read(change_dir)

        now = datetime.utcnow().isoformat() + "Z"
        state = TaskState(
            change_id=change_id,
            repo_path=repo_path,
            status=PLANNING,
            created_at=now,
            updated_at=now,
        )
        change_dir.mkdir(parents=True, exist_ok=True)
        self.write(change_dir, state)
        return state

    def write(self, change_dir: Path, state: TaskState) -> None:
        """Atomic write via tmp + rename."""
        change_dir.mkdir(parents=True, exist_ok=True)
        path = self.state_path(change_dir)
        payload = json.dumps(_state_to_dict(state), indent=2, ensure_ascii=False) + "\n"

        fd, tmp = tempfile.mkstemp(dir=str(change_dir), suffix=".tmp")
        try:
            os.write(fd, payload.encode("utf-8"))
            os.fsync(fd)
            os.close(fd)
            os.rename(tmp, str(path))
        except BaseException:
            try:
                os.close(fd)
            except OSError:
                pass
            try:
                os.unlink(tmp)
            except OSError:
                pass
            raise

    def update(
        self,
        change_dir: Path,
        fn: Callable[[TaskState], TaskState],
    ) -> TaskState:
        """Read → mutate → write (CAS-style, single-process).

        The *fn* callback receives the current state and must return the
        updated state.  ``updated_at`` is refreshed automatically.
        """
        state = self.read(change_dir)
        updated = fn(state)
        updated.updated_at = datetime.utcnow().isoformat() + "Z"
        self.write(change_dir, updated)
        return updated

    def transition(
        self,
        change_dir: Path,
        expected_status: str,
        new_status: str,
    ) -> TaskState:
        """Convenience CAS transition that validates *expected_status*."""
        def _apply(s: TaskState) -> TaskState:
            if s.status != expected_status:
                raise ValueError(
                    f"Status mismatch: expected {expected_status}, got {s.status}"
                )
            s.status = new_status
            return s
        return self.update(change_dir, _apply)

    # ------ digest ------

    def compute_plan_digest(self, change_dir: Path) -> str:
        """SHA-256 digest over the plan artefacts (matches TypeScript version)."""
        h = hashlib.sha256()
        workspace = change_dir.parent.parent.parent  # openspec/changes/<id> -> workspace

        files = [
            workspace / "openspec" / "config.yaml",
            change_dir / "proposal.md",
            change_dir / "design.md",
            change_dir / "tasks.md",
        ]

        for f in files:
            content = ""
            if f.exists():
                content = f.read_text(encoding="utf-8")
            content = content.replace("\r\n", "\n").rstrip() + "\n"
            h.update(str(f).encode("utf-8"))
            h.update(b"\n")
            h.update(content.encode("utf-8"))
            h.update(b"\n\n")

        return h.hexdigest()

    # ------ queries ------

    def list_active(self, workspace: Path) -> list[TaskState]:
        """List all changes with non-terminal status."""
        changes_dir = workspace / "openspec" / "changes"
        if not changes_dir.exists():
            return []

        results: list[TaskState] = []
        for child in changes_dir.iterdir():
            if not child.is_dir():
                continue
            state_file = self.state_path(child)
            if not state_file.exists():
                continue
            try:
                state = self.read(child)
                if state.status not in TERMINAL_STATUSES:
                    results.append(state)
            except Exception as e:
                logger.warning(f"Failed to read state in {child}: {e}")
        return results

    # ------ logging ------

    def log(self, change_dir: Path, run_id: str | None, message: str) -> None:
        """Append a line to progress.log."""
        path = self.progress_log_path(change_dir)
        change_dir.mkdir(parents=True, exist_ok=True)
        ts = datetime.utcnow().isoformat() + "Z"
        line = f"[{ts}] runId={run_id or '-'} {message}\n"
        with open(path, "a", encoding="utf-8") as f:
            f.write(line)
